firmware/arduino/src/PiCommunication/PiCommunication.cpp

```diff
@@ -1,112 +1,125 @@
-    return 1;
+    Serial.print("connected\n");
+    return NO_ERROR;
   }
   else {
+    Serial.print("wrongmsg\n");
     return PI_SENT_WRONG_CODE_ERROR;
   }
 }
 
 int PiCommunication::isTimeToSendDataToPi() {
-  unsigned long timeDifference = _lastSentDataTime - millis();
-  if (timeDifference > _timeBetweenPiSending) {
-    return 1;
-  }
-  else {
-    return 0;
-  }
+  return isTime(_lastSentDataTime, _timeBetweenPiSending);
 }
 
 int PiCommunication::isDataAvailable() {
   return Serial.available();
 }
 
-String PiCommunication::getData() {
+String PiCommunication::getDataFromPi() {
   String receivedString = Serial.readStringUntil('\n'); // reads up-to-but-not-including '\n' char
-  tellPiThatWeGotData();
+  tellPiThatWeGotParameters();
   return receivedString;
 }
 
-void PiCommunication::tellPiThatWeGotData() {
+void PiCommunication::tellPiThatWeGotParameters() {
   Serial.print("G\n");
 }
 
-void PiCommunication::sendServosNotConnectedErrorToPi() {
-  // timeout error
-}
-
-void PiCommunication::sendDataToPi(Data data, State state) {
-  // Send checksum (XOR)
-  // Send running average pressure (signed integer)
-  // Send running average flow
-  // Send “G” for good or “E**” for error and error number
-  // Send “\n”
-  unsigned char checkSum = 0;
-  unsigned char averagePressure = (unsigned char) data.pressureSum / data.numPressureMeasurements;
-  unsigned char batteryPercentage = (unsigned char) data.batteryPercentage;
-  unsigned char breathCompleted = (unsigned char) state.isStartingNewBreath;
-  short flowIntegral;
-  if (state.isStartingNewBreath) {
-    flowIntegral = (short) floor(LITERS_TO_MILLILITERS*data.flowIntegral); // mL/min
+void PiCommunication::sendDataToPi(Data &data, State &state) {
+  /*
+  What we send:
+  - Checksum (XOR) (8 bits)
+  - Battery percentage (8 bit unsigned integer)
+  - Breath complete (8 bit unsigned integer): 0 or 1
+  - Tidal Volume in mL (32 bit unsigned integer). Only use this value if breath complete == 1.
+  - Error code (8 bit unsigned integer). If 0, it’s fine.
+  - Average pressure
+  - O2
+  - Send “\n”
+
+  */
+  uint8_t checkSum = 0;
+  uint8_t batteryPercentage = (uint8_t) data.batteryPercentage;
+  uint8_t breathCompleted = (uint8_t) state.breathCompleted;
+  uint32_t tidalVolume;
+  if (breathCompleted) {
+    tidalVolume = (uint32_t) round(LITERS_TO_MILLILITERS*data.tidalVolume); // mL/min
   } else {
-    flowIntegral = 0;
+    tidalVolume = 0;
   }
-  int error = "G"; // TODO: make better
+
+  uint8_t abnormalPressure = 0; // TODO: get it
+  uint8_t abnormalFiO2 = 0; // TODO: data.fiO2
+  
+  uint8_t errorCode = NO_ERROR; // TODO: actual error maybe state.error?
   
   Serial.write(checkSum);
-  Serial.write(averagePressure);
   Serial.write(batteryPercentage);
   Serial.write(breathCompleted);
-  Serial.write(flowIntegral);
-  Serial.write(error);
+  Serial.write(tidalVolume);
+  Serial.write(errorCode);
+  Serial.write(abnormalPressure);
+  Serial.write(abnormalFiO2);

```

Comment:
do we want to send abnormal only or do we want to send always and use errorcode to determine abnormal on the pi side?